composer/trainer/trainer.py

```diff
@@ -140,6 +246,140 @@ def _unpack_evaluators(
     return evaluators
 
 
+def _unpack_grad_accum(grad_accum: Union[int, str], device: Device) -> Tuple[int, bool]:
+    # Set initial grad_accum to 1 if using adaptive
+    adaptive_grad_accum = grad_accum == "auto"
+    if adaptive_grad_accum:
+        grad_accum = 1
+        warnings.warn(textwrap.dedent("""Setting `grad_accum='auto'` is an experimental feature
+            which may cause uncaught Cuda Out of Memory errors. In this case, please manually
+            set grad_accum explicitly to an integer instead.
+            """),
+                      category=UserWarning)
+    # Cannot use adaptive grad accum on CPU
+    if isinstance(device, DeviceCPU) and adaptive_grad_accum:
+        raise ValueError("Cannot use adaptive grad_accum on CPU. Please set grad_accum >= 1")
+    # grad_accum should be int as we've already handled "auto" case
+    if isinstance(grad_accum, str):
+        raise ValueError("grad_accum must be an int or ``auto``")
```

Comment:
This doesn't actually enforce that grad_accum is an `int`. e.g. I could have passed in `grad_accum=1.5` and that would have passed this check. Instead, we want the error message to match the check. Suggest moving this guard earlier to something like..
```
if grad_accum == "auto":
    return (1, True)  # grad_accum, adaptive_grad_accum
if not isinstance(grad_accum, int):
    raise ValueError(...)
return (grad_accum, False)
```

thoughts?

cc: @mvpatel2000

Reply:
Passing in anything other than a float or a string would not pass type checking. PyRight infers this automatically and types the return value as an (int, bool) implicitely. I can add an explicit type check, though, in case if the user passed in something else.

Reply:
That is important, we cannot assume the user is using pyright.

Reply:
OK, I will add the check here. But FYI -- I don't think we do explicit type checks anywhere else in the codebase to validate that the user passed in a type that matches the typing annotation. (This is generally undefined behavior -- it may work, it could fail with an exception later, or it could silently work but do something different than expected).

Reply:
Done